metals/src/main/scala/scala/meta/internal/metals/BuildServerConnection.scala

```diff
@@ -18,77 +17,134 @@ import scala.concurrent.Future
+          s"build shutdown: ${currentConnection.get.displayName}",
           e
         )
     }
   }
 
-  private def register[T](e: CompletableFuture[T]): CompletableFuture[T] = {
-    ongoingRequests.add(
-      Cancelable(() => Try(e.completeExceptionally(new InterruptedException())))
-    )
-    e
+  private def reconnect() = {
+    synchronized {
+      Option(awaitingConnection.get) match {
+        case Some(promise) => promise.future
+        case None =>
+          val params = Messages.DisconnectedServer.params()
+          languageClient.showMessageRequest(params).asScala.flatMap {
+            case response
+                if response == Messages.DisconnectedServer.reconnect =>
+              val newPromise = Promise[LauncherConnection]()
+              awaitingConnection.set(newPromise)
+              val reconnect = restablishConnection().map { launcherConnection =>
+                currentConnection.set(launcherConnection)
+                awaitingConnection.set(null)
+                ongoingRequests.addAll(launcherConnection.cancelables)
+                launcherConnection
+              }
+              newPromise.completeWith(reconnect)
+              newPromise.future
+            case _ => throw new InterruptedException
+          }
+
+      }
+    }
+  }
+  private def register[T](
+      action: MetalsBuildServer => CompletableFuture[T]
+  ): CompletableFuture[T] = {
+    val connect = if (currentConnection.get.socketConnection.isClosed) {
+      reconnect()
+    } else {
+      Future.successful(currentConnection.get)
+    }
+    connect
+      .flatMap { launcherConnection =>
+        val e = action(launcherConnection.server)
+        ongoingRequests.add(
+          Cancelable(() =>
+            Try(e.completeExceptionally(new InterruptedException()))
+          )
+        )
+        e.asScala
+      }
+      .recoverWith {
+        case _: IOException =>
+          reconnect().flatMap(conn => action(conn.server).asScala)
+      }
+      .asJava
   }
 
   def compile(params: CompileParams): CompletableFuture[CompileResult] = {
-    register(server.buildTargetCompile(params))
+    register { server =>
+      server.buildTargetCompile(params)
+    }
   }
 
   def mainClasses(
       params: ScalaMainClassesParams
   ): CompletableFuture[ScalaMainClassesResult] = {
-    register(server.buildTargetScalaMainClasses(params))
+    register(server => server.buildTargetScalaMainClasses(params))
   }
 
   def testClasses(
       params: ScalaTestClassesParams
   ): CompletableFuture[ScalaTestClassesResult] = {
-    register(server.buildTargetScalaTestClasses(params))
+    register(server => server.buildTargetScalaTestClasses(params))
   }
 
   def startDebugSession(params: DebugSessionParams): CompletableFuture[URI] = {
-    register(
+    register(server =>
       server
         .startDebugSession(params)
         .thenApply(address => URI.create(address.getUri))
     )
   }
 
+  // TODO it should be FUTURe
```

Comment:
Forgot about this one :fearful: Will fix this and figure out the test failures.

Reply:
Should be fixed - all the methods that were using server now by default try to reconnect if connection is dead.